### PR TITLE
Implement Grand Exchange

### PR DESCRIPTION
**Pull request by Moosheplusplus**
Added support for the Grand Exchange per request of @AnthonyOSX.

### CODE REVIEW

**Comment by Timer** [comment]
Pinging some other developers, how do you guys feel about the creation of an entirely new event for this?
Should it be done in this fashion?

@AnthonyOSX @ArtificialPB @Strikegently @Logicail

**Comment by AnthonyOSX** [comment]
I think an entirely new event for what appears to be a concise MessageListener seems a bit overkill. A new event ID for a Grand Exchange event could be added so users could just check the event ID and handle the message as they please.

File: rt4/GrandExchange.java [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+import org.powerbot.script.Condition;
+import org.powerbot.script.MessageEvent;
+import org.powerbot.script.MessageListener;
+import org.powerbot.script.Random;
+import org.powerbot.script.rt4.*;
+import org.powerbot.script.rt4.Game.Crosshair;
+
+public class GrandExchange extends ClientAccessor implements MessageListener {
+	
+	public static final int[] GE_CLERK = {2148, 2149, 2150, 2151};
```
Comment (AnthonyOSX) [inline rt4/GrandExchange.java]: Determining exchange NPC's should probably be done by using actions instead of static ID's. Future proof :)

Comment (ArtificialPB) [inline rt4/GrandExchange.java]: Finding them by name would be better. There are other npcs in game with "Exchange" action (Tool Leprechauns, Void Knights, etc...). Npc and object ids change after every update in OSRS.

File: rt6/GrandExchange.java [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+import org.powerbot.script.MessageEvent;
+import org.powerbot.script.MessageListener;
+import org.powerbot.script.Random;
+import org.powerbot.script.rt6.*;
+import org.powerbot.script.rt6.Item;
+import org.powerbot.script.rt6.Game.Crosshair;
+
+public class GrandExchange extends ClientAccessor implements MessageListener {
+	
+	public static final int[] GE_CLERK = {1419, 2593, 2240, 2241};
```
Comment (ArtificialPB) [inline rt6/GrandExchange.java]: You're missing some ids: 15529, 15530, 19911, 19912, 19913, 19914, 19915

**Comment by ArtificialPB** [comment]
If https://github.com/powerbot/powerbot/issues/847 gets implemented you could completely ditch the GeEvent/MessageEvent implementation and provide a query for offers.

File: rt4/GrandExchange.java (hunk 1 above)
Comment (AnthonyOSX) [inline rt4/GrandExchange.java]: An action check would be much more generic.

**Comment by Strikegently** [comment]
Unless we get the GE hooks for more complete information, this is fine.

**Comment by Timer** [comment]
I personally don't like the event-based notifications... maybe we should look into doing #847 if developers feel they really want a Grand Exchange API.

**Comment by Moosheplusplus** [comment]
@Timer it would be a better solution than using a MessageListener to dispatch events

**Comment by rs-sliske** [comment]
ge hooks would be better than message listener

the buy methods need to find the item by id, not name, there are too many name collisions, and the ge sorts alphabetically, not by best match (even if you type the name exactly it wont always be the first item in the list)

im working on version that uses id atm.

**Comment by Moosheplusplus** [comment]
Is everyone okay with just removing GeEvent and implement the hook at another time, so people can still use the interface for the grand exchange? That way this can be merged and be ready for v7025?

**Comment by AnthonyOSX** [comment]
@Moosheplusplus yeah remove GE event. Anyone that needs it for now can just use the message handler.

**Comment by ArtificialPB** [comment]
Fine with me @Moosheplusplus

**Comment by swatarianess** [comment]
I have the Grand-Exchange hook if anyone wants it. I have had it for almost a month.

**Comment by AnthonyOSX** [comment]
@swatarianess your hook is only applicable for the current game revision. One of the main developers needs to hook it within the updater.

**Comment by Moosheplusplus** [comment]
@TImer @Paris This Pull Request is ready for v7026 and does not require any hooks. This will serve as an interface for the Grand Exchange, until the proper hooks have been implemented.

**Comment by Strikegently** [comment]
If we're going to do it, we should do it right the first time rather than coming back to clean up.
We can just postpone a Grand Exchange API until we get the hooks.

@Timer

**Pull request closed without merging**